src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+            {
+                Debug.Assert(lockEntry._readerLevel == 0);
+            }
+            // Check for nested reader
+            else if (lockEntry._readerLevel != 0)
+            {
+                Debug.Assert((_state & LockStates.ReadersMask) != 0);
+
+                if (lockEntry._readerLevel == MaxAcquireCount)
+                {
+                    throw new OverflowException(SR.Overflow_UInt16);
+                }
+                ++lockEntry._readerLevel;
+                return;
+            }
+            // Check if the thread already has writer lock
+            else if (_writerID == GetCurrentThreadID())
+            {
+                AcquireWriterLock(millisecondsTimeout);
+                Debug.Assert(lockEntry.IsEmpty);
+                return;
+            }
+            else
+            {
+                int spinCount = 0;
+                int currentState = _state;
+                do
+                {
+                    int knownState = currentState;
+
+                    // Reader need not wait if there are only readers and no writer
+                    if (knownState < LockStates.ReadersMask ||
+                        (
+                            (knownState & LockStates.ReaderSignaled) != 0 &&
+                            (knownState & LockStates.Writer) == 0 &&
+                            (
+                                // A waiting reader, after successfully completing the wait, expects that it can become a
+                                // reader, so ensure that there is enough room for waiting readers and this potential reader.
+                                (
+                                    (knownState & LockStates.ReadersMask) +
+                                    ((knownState & LockStates.WaitingReadersMask) >> LockStates.WaitingReadersShift)
+                                ) <= LockStates.ReadersMask - 2
+                            )
+                        ))
+                    {
+                        // Add to readers
+                        currentState = Interlocked.CompareExchange(ref _state, knownState + LockStates.Reader, knownState);
+                        if (currentState == knownState)
+                        {
+                            // One more reader
+                            break;
+                        }
+                        continue;
+                    }
+
+                    // Check for too many readers or waiting readers, or if signaling is in progress
+                    if ((knownState & LockStates.ReadersMask) == LockStates.ReadersMask ||
+                        (knownState & LockStates.WaitingReadersMask) == LockStates.WaitingReadersMask ||
+                        (knownState & LockStates.CachingEvents) == LockStates.ReaderSignaled)
+                    {
+                        // Sleep for a while, then update to the latest state and try again
+                        Helpers.Sleep(1000);
```

Comment:
Changed the sleep duration to 100 ms for the third case. There could be many waiting readers and it could take a while for them to get processor time to acquire the read lock. On the other hand, this happens every time a write lock is released, and when there are only a few waiting readers, any new read request that happens to check the state at the wrong time would block for a second. This duration also bypasses the requested timeout. 100 ms is probably a decent compromise.